crates/wysiwyg/src/composer_model/quotes.rs

```diff
@@ -0,0 +1,319 @@
+// Copyright 2022 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::dom::nodes::dom_node::DomNodeKind::{Generic, Quote};
+use crate::{
+    ComposerAction, ComposerModel, ComposerUpdate, DomHandle, DomNode,
+    UnicodeString,
+};
+
+impl<S> ComposerModel<S>
+where
+    S: UnicodeString,
+{
+    pub fn quote(&mut self) -> ComposerUpdate<S> {
+        if self.action_is_reversed(ComposerAction::Quote) {
+            self.remove_quote()
+        } else {
+            self.add_quote()
+        }
+    }
+
+    fn add_quote(&mut self) -> ComposerUpdate<S> {
+        let (s, e) = self.safe_selection();
+        let Some(wrap_result) = self.state.dom.find_nodes_to_wrap_in_block(s, e) else {
+            // No suitable nodes found to be wrapped inside the quote. The Dom should be empty
+            self.state.dom.append_at_end_of_document(DomNode::new_quote(vec![DomNode::new_text(S::zwsp())]));
+            self.state.start += 1;
+            self.state.end += 1;
+            return self.create_update_replace_all();
+        };
+
+        let parent_handle = wrap_result.ancestor_handle;
+
+        let start_handle = wrap_result.start_handle;
+        let end_handle = wrap_result.end_handle;
+
+        let mut subtree = self.state.dom.split_sub_tree_between(
+            &start_handle,
+            0,
+            &end_handle,
+            usize::MAX,
+            parent_handle.depth(),
+        );
+        // Needed to be able to add children
+        subtree.set_handle(DomHandle::root());
+
+        let start_handle_is_start_at_depth =
+            start_handle.raw().iter().all(|i| *i == 0);
+        let mut insert_at_handle =
+            if subtree.is_block_node() && subtree.kind() != Generic {
+                start_handle.sub_handle_up_to(parent_handle.depth())
+            } else {
+                start_handle.sub_handle_up_to(parent_handle.depth() + 1)
+            };
+        if !start_handle_is_start_at_depth
+            && self.state.dom.contains(&insert_at_handle)
+        {
+            insert_at_handle = insert_at_handle.next_sibling();
+        } else if self.state.dom.document().children().is_empty() {
+            insert_at_handle = self.state.dom.document_handle().child_handle(0);
+        }
```

Comment:
similiar code already seen in code blocks... maybe we can make a function out of it?